### PR TITLE
YUNIKORN-645. Fixed status code issues, error handling issues etc

### PR DESCRIPTION
**Pull request by manirajv06**
### What is this PR for?
A few sentences describing the overall goals of the pull request's commits.
First time? Check out the contributing guide - http://yunikorn.apache.org/community/how_to_contribute   

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/YUNIKORN/
* Put link here, and add [YUNIKORN-*Jira number*] in PR title, eg. `[YUNIKORN-2] Gang scheduling interface parameters`

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/webservice/webservice.go [hunk 1]
```diff
@@ -81,7 +82,7 @@ type YResponseWriter struct {
 
 func (rw *YResponseWriter) Write(bytes []byte) (int, error) {
 	rw.body = bytes
-	if rw.statusCode == -1 {
+	if rw.statusCode == -1 || (rw.statusCode >= http.StatusOK && rw.statusCode <= http.StatusIMUsed) {
```
Comment (yangwwei) [inline pkg/webservice/webservice.go:85]: I still have concerns about checking the status code. This doesn't seem to be a reliable solution. I suggest reverting the wrapper change and continue to use the default writer, which is safer for me. If you look at https://golang.org/pkg/net/http/#Client.Get, the go client library will return an error with some internal logic and check over the HTTP status code. If the wrapper on the server-side wrongly operates on responses, we are in trouble.

**Review by yangwwei — CHANGES_REQUESTED** [review]
hi @manirajv06 

Pls see my comments. I have concerns that the wrapper may cause more problems. To solve the original problem, can we use this approach:

```
// ToJson() wraps the error messages to JSON format
http.Error(w, utils.ToJson(err), http.StatusInternalServerError)
```

I think this can solve the issue and also a safer approach. Let me know if this makes sense.

**Review by wilfred-s — APPROVED** [review]
LGTM

**Review by kingamarton — CHANGES_REQUESTED** [review]
The fix itself LGTM, I added one test case in the following PR: https://github.com/apache/incubator-yunikorn-core/pull/267/files
Please add that UT to your change, so this bug will be covered as well.

**Comment by manirajv06** [comment]
> The fix itself LGTM, I added one test case in the following PR: https://github.com/apache/incubator-yunikorn-core/pull/267/files
> Please add that UT to your change, so this bug will be covered as well.

Taken care

**Comment by yangwwei** [comment]
Cool, thank you  @manirajv06.
Can you please fix the lint issue: 
```
pkg/webservice/handlers_test.go:43: File is not `goimports`-ed with -local github.com/apache/incubator-yunikorn (goimports)
	"github.com/prometheus/client_golang/prometheus/promhttp"
make: *** [Makefile:59: lint] Error 1
Error: Process completed with exit code 2.
```

File: pkg/webservice/handlers.go [hunk 2]
```diff
@@ -211,6 +211,14 @@ func writeHeaders(w http.ResponseWriter) {
 	w.Header().Set("Access-Control-Allow-Headers", "X-Requested-With,Content-Type,Accept,Origin")
 }
 
+func buildJSONErrorResponse(w http.ResponseWriter, detail string, code int) {
+	w.WriteHeader(code)
+	errorInfo := dao.NewYAPIError(nil, code, detail)
+	if jsonErr := json.NewEncoder(w).Encode(errorInfo); jsonErr != nil {
```
Comment (yangwwei) [inline pkg/webservice/handlers.go:217]: One doubt here: I think this line will write the response to the client with a successful code, correct?
how can we make sure the client-side can get the error code? Having the error code encoded in the response body is not enough.

**Review by yangwwei — CHANGES_REQUESTED** [review]

File: pkg/webservice/handlers.go (hunk 2 above)
Comment (manirajv06) [reply pkg/webservice/handlers.go:217]: No, @yangwwei. We are writing the HTTP status code as part of response headers as well. Actual response body has been generated based on the following struct:

	StatusCode  int    `json:"status_code"`
	Message     string `json:"message"`
	Description string `json:"description"`

In general, client reads the status code from response headers and takes decision accordingly. I assuming this should be the case even with our UI component.

Comment (yangwwei) [reply pkg/webservice/handlers.go:217]: Oh, misread that part. Thanks for the explaination.

**Review by yangwwei — APPROVED** [review]
LGTM

**Pull request merged**